edge-hub/src/Microsoft.Azure.Devices.Routing.Core/endpoints/StoringAsyncEndpointExecutor.cs

```diff
@@ -132,63 +132,153 @@ public async Task SetEndpoint(Endpoint newEndpoint, IList<uint> priorities)
             }
         }
 
-        void UpdatePriorities(IList<uint> priorities)
+        public async Task UpdatePriorities(IList<uint> priorities, Option<Endpoint> newEndpoint)
         {
+            Preconditions.CheckArgument(priorities.Count > 0);
+
+            if (this.closed)
+            {
+                throw new InvalidOperationException($"Endpoint executor for endpoint {this.Endpoint} is closed.");
+            }
+
             // Update priorities by merging the new ones with the existing.
             // We don't ever remove stale priorities, otherwise stored messages
             // pending for a removed priority will never get sent.
-            IList<uint> snapshot = this.priorities.Value;
-            this.priorities.Value = priorities
-                .Union(snapshot)
-                .OrderBy(p => p)
-                .ToList();
+            OrderedDictionary snapshot = this.prioritiesToFsms;
+            foreach (uint priority in priorities)
+            {
+                if (!snapshot.Contains(priority))
+                {
+                    string id = GetMessageQueueId(this.Endpoint.Id, priority);
+
+                    // Create a message queue in the store for every priority we have
+                    await this.messageStore.AddEndpoint(id);
+
+                    // Create a checkpointer and a FSM for every message queue
+                    ICheckpointer checkpointer = await this.checkpointerFactory.CreateAsync(id);
+                    EndpointExecutorFsm fsm = new EndpointExecutorFsm(this.Endpoint, checkpointer, this.config);
+
+                    // Add it to our dictionary
+                    snapshot.Add(priority, fsm);
+                }
+                else
+                {
+                    // Update the existing FSM with the new endpoint
+                    var fsm = (EndpointExecutorFsm)snapshot[priority];
+                    newEndpoint.ForEach(async e => await fsm.RunAsync(Commands.UpdateEndpoint(e)));

```

Comment:
does this need to be `ForEachAsync` and awaited:

```suggestion
                    await newEndpoint.ForEachAsync(e => fsm.RunAsync(Commands.UpdateEndpoint(e)));

```

Reply:
is there a difference at runtime? O_o

---
In reply to: [392289666](https://github.com/Azure/iotedge/pull/2670#discussion_r392289666) [](ancestors = 392289666)

Reply:
talked offline, I should be awaiting on a ForEachAsync

---
In reply to: [392457843](https://github.com/Azure/iotedge/pull/2670#discussion_r392457843) [](ancestors = 392457843,392289666)